File: backend/app/models/character_map.py
from typing import Literal, Optional
from pydantic import BaseModel


class Faction(BaseModel):
    id: str
    label: str
    description: str
    color_hint: Literal["blue", "red", "green", "amber", "violet", "slate"]


class ActorInfo(BaseModel):
    name: str
    tmdb_person_id: int
    headshot_url: Optional[str] = None  # null when TMDB has no profile photo


class CreatorInfo(BaseModel):
    # 'author' for books, 'director' for film/tv. Populated post-LLM from
    # OpenLibrary (author name only) or TMDB credits (director with person id).
    kind: Literal["author", "director"]
    name: str
    tmdb_person_id: Optional[int] = None
    headshot_url: Optional[str] = None


class Character(BaseModel):
    id: str
    name: str
    role: str
    description: str
    faction_id: Optional[str]
    importance: Literal["protagonist", "major", "supporting", "minor"]
    is_deceased_in_work: bool
    # Optional so pipeline can detect and default missing values to 3
    spoiler_level: Optional[Literal[0, 1, 2, 3]] = None
    # Short grounding phrase emitted by the LLM to ensure each name is tied to
    # a specific scene or relationship in the source. Optional during rollout.
    name_evidence: Optional[str] = None
    actor: Optional[ActorInfo] = None
    # Coarse geographic origin (region / city / holdfast) for geographic-view
    # rendering. Free-form string — the frontend resolves it against a per-work
    # region→coordinate table. None for works without a defined geography.
    home_region: Optional[str] = None
    # True only for narrative viewpoint (POV) characters — chapters/sections
    # told from their perspective (e.g. ASOIAF POV chapters). Populated by
    # Stage 2 from the analysis's Viewpoint section; default False.
    is_pov: bool = False


class Relationship(BaseModel):
    from_id: str
    to_id: str
    type: Literal[
        "alliance", "family", "romantic", "antagonism",
        "professional", "mentorship", "criminal",
    ]
    label: str
    # Optional so pipeline can detect and default missing values to 3
    spoiler_level: Optional[Literal[0, 1, 2, 3]] = None


class CharacterMap(BaseModel):
    title: str
    subtitle: str
    blurb: str
    spoiler_mode: Literal["full"]
    setting_preamble: Optional[str] = None
    factions: list[Faction]
    characters: list[Character]
    relationships: list[Relationship]
    coverage_note: Optional[str] = None
    # 1-2 sentence summary of substantive adaptation divergences (book vs film,
    # original vs remake). Populated by Stage 2 structuring when the Stage 1
    # analysis surfaces a "Key Adaptation Differences" section.
    adaptation_note: Optional[str] = None
    notes: str
    creator: Optional[CreatorInfo] = None
    # Deep link back to the source the work was resolved from
    # (TMDB movie/tv page or OpenLibrary work page). Populated post-LLM.
    source_url: Optional[str] = None


class RefusalResponse(BaseModel):
    refusal: str

File: frontend/src/types/characterMap.ts
export type ColorHint = 'blue' | 'red' | 'green' | 'amber' | 'violet' | 'slate'
export type Importance = 'protagonist' | 'major' | 'supporting' | 'minor'
export type RelationshipType =
  | 'alliance' | 'family' | 'romantic' | 'antagonism'
  | 'professional' | 'mentorship' | 'criminal'
export type SpoilerLevel = 0 | 1 | 2 | 3

export interface Faction {
  id: string
  label: string
  description: string
  color_hint: ColorHint
}

export interface ActorInfo {
  name: string
  tmdb_person_id: number
  headshot_url: string | null
}

export interface CreatorInfo {
  kind: 'author' | 'director'
  name: string
  tmdb_person_id?: number | null
  headshot_url?: string | null
}

export interface Character {
  id: string
  name: string
  role: string
  description: string
  faction_id: string | null
  importance: Importance
  is_deceased_in_work: boolean
  spoiler_level: SpoilerLevel | null
  actor?: ActorInfo
  home_region?: string | null
  is_pov?: boolean
}

export interface Relationship {
  from_id: string
  to_id: string
  type: RelationshipType
  label: string
  spoiler_level: SpoilerLevel | null
}

export interface CharacterMap {
  title: string
  subtitle: string
  blurb: string
  spoiler_mode: 'full'
  setting_preamble?: string
  factions: Faction[]
  characters: Character[]
  relationships: Relationship[]
  coverage_note?: string
  notes: string
  creator?: CreatorInfo | null
  source_url?: string | null
}
